File: lab 7/scala/src/main/scala/StarvingPhilosopher.scala
class StarvingPhilosopher(val number: Int, val f1: Fork, val f2: Fork, val table: Table)
  extends Philosopher(number, f1, f2)
{

  @throws[InterruptedException]
  override protected def eat(): Unit = {
    table synchronized
    {
      while ( {
        f1.isTaken || f2.isTaken
      }) {
        table.waits = true
        table.wait()
      }
      consume(f1, f2)
      if (table.waits) {
        table.waits = false
        table.notifyAll()
      }
    }
  }
}

File: lab 7/java/src/StarvingPhilosopher.java
public class StarvingPhilosopher extends Philosopher
{
    Table table;
    public StarvingPhilosopher(int number, Fork f1, Fork f2, Table table)
    {
        this.number = number;
        this.f1 = f1;
        this.f2 = f2;
        this.table = table;
    }

    protected void eat() throws InterruptedException
    {
        synchronized (table)
        {
            while(f1.isTaken() || f2.isTaken())
            {
                table.waits = true;
                table.wait();
            }
            consume(f1,f2);
            if(table.waits)
            {
                table.waits = false;
                table.notifyAll();
            }
        }
    }
}
